### PR TITLE
Suggested updates to Ethereum bridge spec

### PR DESCRIPTION
**Pull request by james-chf**
* update `TransferFromEthereum` data type and other pseudocode
* reference the fact that we will use multitoken balances for the wrapped ETH/ERC20

Tried to clarify some other things and there are some TBDs left in which we should design or spec out separate from this PR.

### CODE REVIEW

File: src/architecture/interoperability/ethereum-bridge.md [hunk 1]
```diff
@@ -76,60 +76,50 @@ by at least 2/3 of the staking validators as usual.
...
+    /// height of the Ethereum block at which the message appeared
     height: u64,
-    /// the hash & height of the last descendant block marked as `seen`
+    /// the hash & height of the latest descendant Ethereum block marked as `seen`
     latest_descendant: ([u8; 32], u64)
 }
+```
 
+The minimum number of confirmations indicated in the outgoing Ethereum message
+(maybe defaulting to 25 or 50 if unspecified) specifies the minimum number of
```
Comment (james-chf) [inline src/architecture/interoperability/ethereum-bridge.md:111]: Will we need a default, or could we expect that every TransferFromEthereum message from the Ethereum blockchain would have a min_confirmations explicitly specified?

File: src/architecture/interoperability/ethereum-bridge.md [hunk 2]
```diff
@@ -76,60 +76,50 @@ by at least 2/3 of the staking validators as usual.
...
+confirmations in Ethereum block depth that must be reached before the assets will be
+minted on Namada. This is the purpose of the `/queue` for this validity
+predicate.
+
+The protocol transaction that sets an `/eth_block/$block_hash/seen = true` must also update `/queue` with new transfers from that block. `/eth_block/$block_hash/seen = true` implies that that Ethereum block has been processed by the Ethereum bridge. Every time an `/eth_block/$block_hash/seen` becomes `true`, the following must happen:
+
+1. For each existing `TransferFromEthereum` in the `/queue`, update its number of confirmations. This
```
Comment (james-chf) [inline src/architecture/interoperability/ethereum-bridge.md:118]: My understanding of how the transactions would work here (in the context of ABCI++):

The block proposer constructs a transaction to update `/eth_block` state based on the vote extensions from the last round. This same transaction will update `#EthBridge/queue` (steps 1 and 2 here).

Separately, the block proposer should construct another transaction based on looking for messages in `/queue` which reached the min confirmations in the previous block (step 3 here). This transaction doesn't need to interact with `/eth_block` storage and will work solely with `#EthBridge` storage i.e the `/queue` and multitoken keys.

File: src/architecture/interoperability/ethereum-bridge.md [hunk 3]
```diff
@@ -76,60 +76,50 @@ by at least 2/3 of the staking validators as usual.
...
-        &self.0
-    }
+enum AssetFromEthereum {
+    Eth,  // Native ETH
+    Erc20(EthereumAddress),
+    WrappedNamadaToken(NamadaAddress),
 }
 
-/// Generic struct for transferring value from Ethereum
 struct TransferFromEthereum<Token: MintingAddress> {
```
Comment (batconjurer) [inline src/architecture/interoperability/ethereum-bridge.md:94]: Remove the generic argument here.

File: src/architecture/interoperability/ethereum-bridge.md [hunk 4]
```diff
@@ -228,51 +237,33 @@ Signing incorrect headers is considered a slashable offense. Anyone witnessing
...
 }
 ```
-A special Namada validity predicate will be called on this transaction. If the
-transaction is valid, the corresponding amount of the NAM token will be transferred
-from the `source` address and deposited in an escrow account by the
-validity predicate.
 
-Just as in redeeming ETH above, it is incumbent on the end user to
+If a user wishes to mint a wrapped Namada token on Ethereum, they must submit a transaction on Namada that:
+- stores `MintWrappedNam` on chain somewhere?
```
[the remark this answers is not in the document]
Comment (batconjurer) [reply src/architecture/interoperability/ethereum-bridge.md:259]: I think we just need to escrow the right amount of NAM. We can forget the target address as the value may change hands on Ethereum anyway and when it's redeemed it could go to any address on Namada in theory.

File: src/architecture/interoperability/ethereum-bridge.md (hunk 1 above)
Comment (james-chf) [reply src/architecture/interoperability/ethereum-bridge.md:111]: Should the default be set in the Ethereum smart contracts or un Namada? i.e. can Namada expect every `TransferFromEthereum` it receives from Ethereum to have this `min_confirmations` field already filled in, or will Namada code decide the default to use if this field is empty. cc @Fraccaman not sure which event in the smart contract code corresponds to `TransferFromEthereum` here, guessing [event TransferToNamada](https://github.com/anoma/ethereum-bridge/blob/5fc8519ed5a4b6f3c4972d6c2d58d096c4fd3c47/contracts/interface/IBridge.sol#L7)?

Comment (batconjurer) [reply src/architecture/interoperability/ethereum-bridge.md:111]: I'm happy with making @Fraccaman do the work :P

File: src/architecture/interoperability/ethereum-bridge.md (hunk 2 above)
Comment (james-chf) [reply src/architecture/interoperability/ethereum-bridge.md:118]: :+1: will update this part of the spec. How will a validator update the queue as part of ledger logic and not through a protocol transaction? I'm not sure how to change storage except through a transaction.

File: src/architecture/interoperability/ethereum-bridge.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (james-chf) [reply src/architecture/interoperability/ethereum-bridge.md]: I've added a TBD here for now

File: src/architecture/interoperability/ethereum-bridge.md (hunk 2 above)
Comment (james-chf) [reply src/architecture/interoperability/ethereum-bridge.md]: I've made more explicit about how we will use internal transactions. So an internal transaction (applied in `FinalizeBlock`) will be responsible for steps 1 and 2, and probably step 3 as well. I removed some details about how exactly we'll update the existing number of confirmations for existing `TransferFromEthereum`s as that's TBD but optimistically we can use the method discussed earlier of "incrementing" `latest_descendant` one block at a time as long as we see every Ethereum block (though we will have to have a way to handle also when we have skipped blocks or for forks)

File: src/architecture/interoperability/ethereum-bridge.md [hunk 5]
```diff
@@ -327,6 +328,9 @@ a proof from Namada:
 If all the above verifications succeed, the contract may affect the
 appropriate state change, emit logs, etc.
 
+## Starting the bridge
```
Comment (james-chf) [inline src/architecture/interoperability/ethereum-bridge.md:331]: This needs to be worked out. I think one protocol transaction to initialize all the storage and start the bridge going (maybe by setting the very first `/eth_block/$block_hash`) may work. This could be worked out in another PR.

File: src/architecture/interoperability/ethereum-bridge.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (batconjurer) [reply src/architecture/interoperability/ethereum-bridge.md]: The user should put that in their transaction and I guess we should store it when we escrow.

File: src/architecture/interoperability/ethereum-bridge.md (hunk 5 above)
Comment (batconjurer) [reply src/architecture/interoperability/ethereum-bridge.md:331]: We should do this in the init chain call of the ledger.
